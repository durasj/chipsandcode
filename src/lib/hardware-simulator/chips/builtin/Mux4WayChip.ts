import type { ChipPin } from '../Chip';
import type Chip from '../Chip';

/**
 * Mux4Way chip implemented natively
 */
class Mux4WayChip implements Chip {
  public readonly name = 'Mux4Way' as string;

  private a: boolean[];
  private b: boolean[];
  private c: boolean[];
  private d: boolean[];
  private sel: boolean[];
  private out: boolean[];

  constructor(private width = 1) {
    this.a = new Array(width).fill(false);
    this.b = new Array(width).fill(false);
    this.c = new Array(width).fill(false);
    this.d = new Array(width).fill(false);
    this.sel = [false, false];
    this.out = new Array(width).fill(false);
  }

  setInput(name: 'a' | 'b' | 'c' | 'd' | 'sel', value: boolean[]) {
    this[name] = value;
  }

  getOutput() {
    return this.out;
  }

  getPins() {
    // Constructs map on the fly
    // Pin inspection is less performance sensitive than actual execution

    return new Map([
      ['a', { type: 'input', width: this.width, state: this.a, connections: [] }],
      ['b', { type: 'input', width: this.width, state: this.b, connections: [] }],
      ['c', { type: 'input', width: this.width, state: this.c, connections: [] }],
      ['d', { type: 'input', width: this.width, state: this.d, connections: [] }],
      ['sel', { type: 'input', width: 2, state: this.sel, connections: [] }],
      ['out', { type: 'output', width: this.width, state: this.out, connections: [] }],
    ]) satisfies Map<string, ChipPin>;
  }

  public run() {
    const pin = this.selToPin();
    this.out = pin;
  }

  private selToPin() {
    const sel = this.sel.map((v) => (v ? '1' : '0')).join('');

    switch (sel) {
      case '00':
        return this.a;
      case '01':
        return this.b;
      case '10':
        return this.c;
      case '11':
        return this.d;
      default:
        return this.a;
    }
  }
}

export default Mux4WayChip;
