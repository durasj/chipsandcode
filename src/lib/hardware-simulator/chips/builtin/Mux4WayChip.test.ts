import { describe, expect, it } from 'vitest';

import Mux4WayChip from './Mux4WayChip';

describe('Mux4Way Chip', () => {
  it('Provides interface info', async () => {
    const chip = new Mux4WayChip();

    expect(chip.name).toBe('Mux4Way');
    expect(chip.getPins()).toMatchSnapshot();
  });

  it('Has correct logic', async () => {
    const chip = new Mux4WayChip();

    const values = [
      [[false], [false], [false], [false], [false, false], [false]],
      [[true], [false], [false], [false], [false, false], [true]],
      [[false], [true], [false], [false], [false, true], [true]],
      [[true], [false], [false], [false], [false, true], [false]],
      [[false], [false], [true], [false], [true, false], [true]],
      [[true], [false], [false], [false], [true, false], [false]],
      [[false], [true], [true], [true], [true, true], [true]],
      [[true], [true], [false], [false], [true, true], [false]],
    ];

    values.forEach(([a, b, c, d, sel, out]) => {
      chip.setInput('a', a);
      chip.setInput('b', b);
      chip.setInput('c', c);
      chip.setInput('d', d);
      chip.setInput('sel', sel);
      chip.run();

      expect(chip.getOutput()).toStrictEqual(out);
    });
  });
});
