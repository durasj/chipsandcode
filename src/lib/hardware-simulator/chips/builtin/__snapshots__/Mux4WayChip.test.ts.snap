// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Mux4Way Chip > Provides interface info 1`] = `
Map {
  "a" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "b" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "c" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "d" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "sel" => {
    "connections": [],
    "state": [
      false,
      false,
    ],
    "type": "input",
    "width": 2,
  },
  "out" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "output",
    "width": 1,
  },
}
`;
